csrc/tensor_view.cpp

```diff
@@ -937,6 +937,8 @@ TensorView* TensorView::rFactor(const std::vector<int>& axes) {
   TensorView* producer =
       IrBuilder::create<TensorView>(producer_domain, getDataType().value());
 
+  producer->setDeviceMesh(mesh_);
```

Comment:
Consider the fusion:
```
DeviceMesh mesh({0,1});

TensorView* tv0 = makeConcreteTensor(2);
TensorView* tv1 = sum(tv0, {0});

tv0->setDeviceMesh(mesh);
tv1->setDeviceMesh(mesh);

tv1->split(0, 4);
TensorView* tv2 = tv1->rFactor({0});
```
Then, `tv2`'s device mesh should be set to be the one of `tv1`.

This is why we need this line

Reply:
I think this is fine, but I wonder if we could do some basic inference to automatically set a proper device mesh.

Reply:
Yes, it's not about this PR, but I was wondering if we could do something to make it more robust rather than manually annotating all tenors individually. Think of `parallelizeAllLike`.

Reply:
it turns out we have a function [shardAllLike](https://github.com/NVIDIA/Fuser/blob/b9d3c116aa46dff33bcec3a00fbd2d8c7b2d0ffe/csrc/multidevice/utils.cpp#L115), but I'm not sure how it could help here

Reply:
Nothing urgent. I'm not sure what automation would be the most desirable, but I believe it would become gradually more obvious as we gain more experiences with multi-GPU examples.

Reply:
I now understand it's being used by the manual scheduling test. My concern was that: we have too many scheduling methods creating TensorView (seven occurrences in tensor_view.cpp alone). It's weird to have some but not all of them propagating the device mesh. A bigger question is: at the end of the day, it's unclear to me whether they should be responsible to propagate the device mesh or their users should. 

I'll review that test to see if there's a way to around.

Reply:
I see. In my opinion it depends on the considered function that creates the tensor. In my opinion it makes sense to add this change in factor because:
1) rFactor is really a scheduling tool, that replay the scheduling of the argument to the Tensor it cresates -- and the device mesh is an intrinsic part of that
2) The intention of rFactor is to divide a reduction into several steps, and so we want the intermediate tensor to be sharded on the save device mesh. In other word, in my opinion, in all cases where rFactor is used we will want to set the device mesh this way.

Those two above points might be different for other TensorView constructors.

This being said, it doesn't really matter for me to set manually the rFactor domain in the test. So let me know, if it is a concern let me simply move this to the tests -- it will just result in more verbose tests but that's not a problem for me

Reply:
The current change is good enough! And I don't want to bikeshed too much before we have a clear plan for our Python API. 

I agree with both your points. My concern was about who should be responsible to propagate/infer the device mesh -- rFactor is likely too low-level an interface for **that** purpose, and thus my hesitation.

Reply:
ok so, if that's ok with you, let me resolve this conversation and keep this point as a future discussion